### PR TITLE
Support HTMLTemplateElement.content for #1065

### PR DESCRIPTION
**Pull request by rickychien**
See also #1065 

@domenic here is my quick patch for supporting template.content, I'm not sure what's the best approach to hack that things. Please take a look!

### CODE REVIEW

**Comment by Sebmaster** [comment]
The [spec](https://html.spec.whatwg.org/multipage/scripting.html#template-contents) definitely requires a lot more details with template elements, especially in regards to cloning so modifying the result of content doesn't modify the template element itself.

**Comment by domenic** [comment]
Thanks for working on this!

My biggest concern here is that I'm pretty sure template.content and template.firstChild should actually be different. The former should be a document fragment, whereas the latter should be the first node nested under the template tag in the source.

It'd be OK to have an incomplete implementation, but we should at least get that right.

For the tests, could you either find appropriate web platform tests, or create new ones in https://github.com/tmpvar/jsdom/tree/master/test/web-platform-tests/to-upstream ?

**Comment by rickychien** [comment]
Hey @domenic,

Unfortunately, there are lots of tests cannot passed web platform tests, so I leaved some passable tests and updated the patch.

**Comment by rickychien** [comment]
I was thinking the spec [1] is a little bit unclear to me and I spent lots of time doing trial and errors on real browser to figure out real behavior behind that. (Or maybe I missed something from spec?)

[1] https://html.spec.whatwg.org/multipage/scripting.html#template-contents

File: lib/jsdom/level2/html.js [hunk 1]
```diff
@@ -2217,7 +2217,29 @@ define('HTMLSourceElement', {
...
+              content.appendChild(child);
+            }
+          }
+        };
+        createContent(content, this);
+        this._content = content;
+      }
+      return this._content;
+    },
+    set content(val) {}
```
Comment (domenic) [inline lib/jsdom/level2/html.js]: No empty setter.

File: test/level2/html.js [hunk 2]
```diff
@@ -16670,6 +16670,33 @@ exports.tests = {
     test.done();
   },
 
+  HTMLTemplateElement01: function(test) {
+    var doc = load('template');
```
Comment (domenic) [inline test/level2/html.js]: Instead of using `load()`, can you just use `jsdom.jsdom()`?

**Comment by domenic** [comment]
This is probably ready to pull when I finally get some free time for a jsdom release. But, I'm curious:

- What did you find unclear about the spec?
- Were you not able to get any of the other tests in https://github.com/w3c/web-platform-tests/tree/master/html/semantics/scripting-1/the-template-element passing?

**Comment by rickychien** [comment]
There are only two additions-to-the-steps-to-clone-a-node tests are able to pass after I checked all tests of the-template-element. Some tests got stuck without any error messages so I guess it probably ran into problem during parsing? 

We need to do more works to support correct behavior for template like as only append template's childNodes into template.content in HTML parsing stage.
(I found that it will create a documentFragment object as its first child after HTML parsing)

After that, we can do more improvement :)

**Comment by rickychien** [comment]
About unclear part, for instance:

1. Let doc be the template element's node document's appropriate template contents owner document.
what's doc mean?
2. Create a DocumentFragment object whose node document is doc.
Could you give me an code example how to do this? It may help me understand it quickly.

thank!

**Comment by rickychien** [comment]
Descriptions between [1] and [2] are a little bit different. Keyword node document appears in [1] but  ownerDocument appears in [2]. Which one is better?

[1] https://html.spec.whatwg.org/multipage/scripting.html#template-contents
[2] http://www.w3.org/TR/html5/scripting-1.html#the-template-element

**Comment by domenic** [comment]
Please don't use the outdated W3C fork when implementing jsdom features. Use http://html.spec.whatwg.org/multipage/ instead.

**Comment by rickychien** [comment]
ok I got it, but description of appropriate template contents owner document is same as outdated one...lol

I've updated my patch to move ```jscs:disable maximumLineLength``` in index.js.

**Comment by domenic** [comment]
OK. To create a new Document without a browsing context, you can do `new core.HTMLDocument({ parsingMode: "html" })`.

But yeah, this is still good to pull in when I have the time to do a release, just clarifying :)

**Comment by domenic** [comment]
Merged as 86ef3543de49deb21d75e07179baf5a45a56c7c3; look for a 7.0.0 soon with this in it! Thanks so much for contributing; this is a great feature to have.

**Comment by rickychien** [comment]
nice!!

**Pull request closed without merging**